pubsub/pubsub_test.go

```diff
@@ -72,3 +121,89 @@ func TestSendReceive(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestLotsOfMessagesAndSubscriptions(t *testing.T) {
+	howManyToSend := int(1e3)
+	ctx := context.Background()
+	dt := &driverTopic{}
+
+	// Make subscriptions and start goroutines to receive from them.
+	var subs []*pubsub.Subscription
+	nSubs := 100
+	var wg sync.WaitGroup
+	for i := 0; i < nSubs; i++ {
+		ds := NewDriverSub()
+		dt.subs = append(dt.subs, ds)
+		s := pubsub.NewSubscription(ctx, ds, pubsub.SubscriptionOptions{})
+		subs = append(subs, s)
+		wg.Add(howManyToSend)
+		go func() {
+			for j := 0; j < howManyToSend; j++ {
+				_, err := s.Receive(ctx)
+				if err != nil {
+					panic(err)
+				}
+				wg.Done()
+			}
+		}()
+	}
+
+	// Send messages.
+	topic := pubsub.NewTopic(ctx, dt, pubsub.TopicOptions{})
+	for i := 0; i < howManyToSend; i++ {
+		m := &pubsub.Message{Body: []byte("user signed up")}
+		if err := topic.Send(ctx, m); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Wait for all the goroutines to finish processing all the messages.
+	wg.Wait()
+
+	// Clean up.
+	if err := topic.Close(); err != nil {
+		t.Fatal(err)
+	}
+	for _, s := range subs {
+		if err := s.Close(); err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+func TestCancelSend(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	ds := NewDriverSub()
+	dt := &driverTopic{
+		subs: []*driverSub{ds},
+	}
+	topic := pubsub.NewTopic(ctx, dt, pubsub.TopicOptions{})
+	m := &pubsub.Message{}
+
+	// Intentionally break the driver subscription by acquiring its semaphore.
+	// Now topic.Send will have to wait for cancellation.
+	<-ds.sem
+
+	cancel()
+	if err := topic.Send(ctx, m); err == nil {
+		t.Error("got nil, want cancellation error")
+	}
+	if err := topic.Close(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestCancelReceive(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	ds := NewDriverSub()
+	s := pubsub.NewSubscription(ctx, ds, pubsub.SubscriptionOptions{})
+	go s.Receive(ctx)
+	cancel()
+	// Without cancellation, this Receive would hang.
```

Comment:
I think you're assuming that the goroutine above starts before the cancel() is executed. Actually, I'm not sure why that goroutine is there.

You could just do 
```
cancel()
err := s.Receive(ctx)
```
For a more complicated test where the Receive is in the middle of working when you cancel it, let's wait for the fake. We'll write it to make that kind of test not time-dependent.

Reply:
Done.